Create Content

adrianhacker-pdx opened the pull request:
Added new document.  The fourth piece of the EVM Internals Series  This completes bounty:
https://gitcoin.co/issue/kauri-io/Content/40/3180
I will make any additional corrections as requested in proofreading.  May I be awarded the bounty for completing this project? 
https://gitcoin.co/adrianhacker-pdx

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+## Introduction
+
+Bitcoin has always been the highest volume and value cryptocurrency in digital money. It runs on blockchain technology. Blockchain is a peer to peer, public, encrypted, trust-less ledger. Script is the name of the scripting language that constructs Bitcoin. Script is a mid to lower capability programming language. Public and private key transacting is the model for signing transactions. Besides t ...
```
```suggestion
Bitcoin has always been the highest volume and value cryptocurrency in digital money, and it runs on blockchain technology. Blockchain is a peer to peer, public, encrypted, trust-less ledger. [Script ](https://en.bitcoin.it/wiki/Script) is the name of the scripting language that constructs Bitcoin. Script is a mid to lower capability programming language. Public and private key transacting is the model for signing transactions. Besides transacting digital assets, there is no extra utility built into Bitcoin. 
```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+# The EVM and High-Level Smart Contract Programming Languages
+
+## Introduction
```
```suggestion

```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+Bitcoin has always been the highest volume and value cryptocurrency in digital money. It runs on blockchain technology. Blockchain is a peer to peer, public, encrypted, trust-less ledger. Script is the name of the scripting language that constructs Bitcoin. Script is a mid to lower capability programming language. Public and private key transacting is the model for signing transactions. Besides t ...
+
+Ethereum Blockchain technology offers limitless use case scenarios. A virtual computer deployed on the Ethereum Blockchain creates many possible use cases. Smart contracts are programs written in a high-level programming language. Think of JavaScript syntax with a name of Solidity or Vyper.  These languages build Smart contract code that compiles for use on the Ethereum Blockchain. This computing ...
```
```suggestion
Ethereum Blockchain technology offers limitless use cases, thanks to its virtual machine. Smart contracts are programs written in a high-level programming language, including [Solidity ](https://solidity.readthedocs.io/) and [Vyper](https://vyper.readthedocs.io/en/latest/installing-vyper.html). These languages build Smart contract code that compiles for use on the Ethereum Blockchain. This computing system revolutionizes privacy, security, and freedom otherwise unrealized in present-day governance.
```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+# The EVM and High-Level Smart Contract Programming Languages
+
+## Introduction
```
I'm not sure how much we need the introduction, this post is an EVM deep-dive, so I think people know what it is.

adrianhacker-pdx replied:
I gutted this and re-wrote 75% of it, please review.  Thanks!

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+## Ethereum Virtual Machine Feature Overview
+
+High-level is a vague description of a computing platform's capability. Precise performance metrics would go beyond the scope of this article. The Ethereum Blockchain is a powerful and isolated virtual computing environment. Computing features of the Ethereum Virtual Machine (EVM) include: 
+
```
Could you add a tiny bit of detail on what these mean, or at least maybe link to the wiki or something like that?

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+* Sand boxing 
+* Two-way interaction with other smart contracts 
+
+Simple transaction protocol initiates the deployment of very powerful applications on the EVM. The Ethereum protocol is trust-less; secured by peer-to-peer consensus and data encryption. It is very difficult, impossible, to access unauthorized information. The program code is verifiable and impossible to alter. Always available to every user as the same version and build. Pushing updates that use ...
```
A lot of this paragraph is a bit basic, and a bit vague. Again, this is a post about an EVM deep dive. A couple of words here and there about what these terms mean is all that's needed.

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+Simple transaction protocol initiates the deployment of very powerful applications on the EVM. The Ethereum protocol is trust-less; secured by peer-to-peer consensus and data encryption. It is very difficult, impossible, to access unauthorized information. The program code is verifiable and impossible to alter. Always available to every user as the same version and build. Pushing updates that use ...
+
+Prebuilt functions are built in the EVM.  Among other things these functions create data abstraction between the user and the actual blockchain implementation.  Direct features of the EVM are listed below that have an aggregate functionality of a self contained computer system.
```
```suggestion
Prebuilt functions are built in the EVM.  Among other things these functions create data abstractions between the developer and the actual blockchain implementation.  Direct features of the EVM are listed below that have an aggregate functionality of a self contained computer system.
```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+Prebuilt functions are built in the EVM.  Among other things these functions create data abstraction between the user and the actual blockchain implementation.  Direct features of the EVM are listed below that have an aggregate functionality of a self contained computer system.
+
+* Simple transaction methods for new smart contract deployment
```
What does this mean?

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+Prebuilt functions are built in the EVM.  Among other things these functions create data abstraction between the user and the actual blockchain implementation.  Direct features of the EVM are listed below that have an aggregate functionality of a self contained computer system.
+
+* Simple transaction methods for new smart contract deployment
+* Smart contract interaction only by address
```
What does this mean?

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+* Simple transaction methods for new smart contract deployment
+* Smart contract interaction only by address
+* Arguments and data pass in from the user and pass out to a user or contract
```
```suggestion
* Argument passing between contracts
```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+* Addressable persistent storage acts as a virtual hard drive 
+* Non-persistent storage acts as system memory
+
+A closer look at the EVM and dApp ecosystem show a complex orchestration of silently running programs and processes. Technology today offers many choices in a programming language presenting obstacles in compatibility. Cross-platform data abstraction requires a complex web of behind the scenes computing. 
```
This paragraph isn't really helpful. Maybe it's better to reword to explain that contracts interact with each other build from each other.

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+## Discrete Parts that make the EVM
+
+As stated before, the EVM is prebuilt with libraries of ready to use functions. Smart contract languages such as Solidity and Vyper are simple, feature-rich languages. Embedded function libraries manipulate data in ways making the EVM quasi-Turing complete. Turing complete system is so named on the basis that it can use a lot of math proving it can solve any algorithm.
```
Again, this doesn't really add anything.

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+<a href="https://imgur.com/J3lKgEy"><img src="https://i.imgur.com/J3lKgEy.png" title="source: imgur.com" /></a>
+
+Opcodes are a large library of functions preprogrammed into the EVM. A list of all Opcodes available is accessible [here](https://www.ethervm.io/#opcodes). Fewer resources handle these functions because they are small, 4-bit hex identifiers. Every function found in JavaScript or Python exists also in the EVM. Opcodes include functions for data manipulation, I/O, storage, security, and logic. High ...
```
```suggestion
Opcodes are a large library of functions preprogrammed into the EVM. You can find a list of all Opcodes [in the documentation](https://www.ethervm.io/#opcodes). Fewer resources handle these functions because they are small, 4-bit hex identifiers. Every function found in JavaScript or Python exists also in the EVM. Opcodes include functions for data manipulation, I/O, storage, security, and logic. High-level smart contract language compiles into bytecode. This is data passed into the EVM in hex that calls opcode functions and handles user arguments. The image located at the top of the page is a matrix of two digit hex opcodes. 
```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+### Transactions
+
+Smart Contracts are simple to deploy on the EVM blockchain. Every type of transaction uses the same format to execute the EVM. Two main smart contact transactions call the contract or deploy a new contract. Either of these transactions are capable of calling a large variety of functions. Functions called can be user defined. Data passes in and out of the EVM in either case. This data can pass bac ...
```
```suggestion
Smart Contracts are deployed on the EVM blockchain. Every type of transaction uses the same format to execute the EVM. Two main smart contact transactions call the contract or deploy a new contract. Either of these transactions are capable of calling a large variety of functions. Functions called can be user defined. Data passes in and out of the EVM in either case. This data can pass back to the user or forward on to another contract.
```

ChrisChinchilla replied:
The EVM is not the blockchain

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+<a href="https://imgur.com/bfKk7mY"><img src="https://i.imgur.com/bfKk7mY.png" title="source: imgur.com" /></a>
+
+The image above shows how a transaction flow can happen in the EVM. Data is passed out of the EVM in what is known as a 'block'. A block is the final result of the arguments, variables, functions, and transactions presented. The data has been manipulated and encrypted. A proof of work puzzle solved by network miners cryptographically proves data integrity. It also cryptographically ties the compl ...
```
```suggestion
The image above shows how a transaction flow happens in the EVM. Data is passed out of the EVM in what is known as a 'block'. A block is the final result of the arguments, variables, functions, and transactions presented. The data has been manipulated and encrypted. A proof of work puzzle solved by network miners cryptographically proves data integrity. It also cryptographically ties the completed block to adjacent data blocks.
```

ChrisChinchilla replied:
What do you mean by "manipulated"? What's proof of work? Few details missing here

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+## Miners - The actual computing power of the EVM
+
+Key features of the Ethereum Blockchain include decentralization and peer to peer networking. A node can be run by any individual with hardware meeting minimum system requirements. Transactions are broadcast on to the network so that multiple nodes can see them at the same time. A fee is included for the miner that completes the proof of work and finishes a block. This spider web type of prolifer ...
```
```suggestion
Key features of the Ethereum Blockchain include decentralization and peer to peer networking. A node can be run by any individual with hardware meeting minimum system requirements. Transactions are broadcast on to the network so that multiple nodes can see them at the same time. A fee is included for the miner that completes the proof of work and finishes a block. This spider web type of proliferation allows the nodes to compare the information they have received and agree that it matches. This comparison is performed among many nodes to create what is called consensus, or agreement that the values now held in block are the true values that represent the current state of the EVM.
```

ChrisChinchilla commented on EVM and High Level Languages.md:
```diff
@@ -0,0 +1,95 @@
+
+The above image shows a visual relationship between the high-level languages that are used for dApp programming and the Ethereum Blockchain. Typical dApp implementations may contain multiple programming languages and protocols. The dApp is therefore often created using a container for development and deployment. A container creates an enclosed running environment that holds utilities, libraries,  ...
+
+NPM, Yarn, Django, and React are all package managers that install and abstract the dependencies required in dApp deployment. These various modules handle everything from coding, test flow, and smart contract deployment. This list is not exhaustive of the features and development flows requiring multiple app dependencies installed. During development, the installed application modules are fine tu ...
```
Django and React are not package managers

ChrisChinchilla commented:
@adrianhacker-pdx I started a review, but to be blunt, this is not as good as your last post. It has inaccuracies, and a lot of content that is not useful or detailed enough in the context of an "EVM deep dive" aimed at people who have a rough idea of what they're doing already.

Not sure if you're happy to make changes, or I can just pay out half the remaining bounty for the other post plus a small tip?

adrianhacker-pdx commented:
@ChrisChinchilla I rewrote a good portion of the article.  I think you will find it more in-depth.  I removed common knowledge items.  Can you tell me how to add header images?  I can't figure it out on Kauri.io.  Thank you very much for this opportunity.

adrianhacker-pdx commented:
P.S.  That rough draft was bad... I was extremely tired, my brain was mush, and I was convinced that my words made sense.  :) :)

adrianhacker-pdx commented:
@ChrisChinchilla Here is a link to the collection that I have curated on Kauri.io with all four articles.  All with header images :).  Let me know if you have any formatting suggestions.  I think it looks nice.  I will update the 4th article after editing to your satisfaction.   https://kauri.io/a-deep-dive-into-the-ethereum-virtual-machine-evm/5e40dd7b26af2700016d5668/c

ChrisChinchilla commented:
@adrianhacker-pdx Ah, you went ahead and posted anyway before this was closed. So I will make suggestions there.
